src/apprt/gtk/portal/OpenURI.zig

```diff
@@ -0,0 +1,400 @@
+const OpenURI = @This();
+
+const std = @import("std");
+const Allocator = std.mem.Allocator;
+
+const gio = @import("gio");
+const glib = @import("glib");
+const gobject = @import("gobject");
+
+const App = @import("../App.zig");
+const portal = @import("../portal.zig");
+const apprt = @import("../../../apprt.zig");
+
+const log = std.log.scoped(.openuri);
+
+/// The GTK app that we "belong" to.
+app: *App,
+
+/// Connection to the D-Bus session bus that we'll use for all of our messaging.
+dbus: *gio.DBusConnection,
+
+/// Mutex to protect modification of the entries map or the cleanup timer.
+mutex: std.Thread.Mutex = .{},
+
+/// Map to store data about any in-flight calls to the portal.
+entries: std.AutoHashMapUnmanaged(usize, Entry) = .empty,
+
+/// Used to manage a timer to clean up any orphan entries in the map.
+cleanup_timer: ?c_uint = null,
+
+/// Data about any in-flight calls to the portal.
+pub const Entry = struct {
+    /// When the request started.
+    start: std.time.Instant,
+    /// A token used by the portal to identify requests and responses. The actual
+    /// format of the token does not really matter as long as it can be used as
+    /// part of a D-Bus object path. `usize` was chosen since it's easy to hash and
+    /// to generate random tokens.
+    token: usize,
+    /// The "kind" of URI. Unused here, but we may need to pass it on to the fallback
+    /// URL opener if the D-Bus method fails.
+    kind: apprt.action.OpenUrl.Kind,
+    /// A copy of the URI that we are opening. We need our own copy since the method
+    /// calls are asynchronous and the original may have been freed by the time we need
+    /// it.
+    uri: [:0]const u8,
+    /// Used to manage a scription to a D-Bus signal, which is how the Portal
+    /// reports results of the method call.
+    subscription: ?c_uint = null,
+
+    pub fn deinit(self: *Entry, alloc: Allocator) void {
+        alloc.free(self.uri);
+    }
+};
+
+pub const Errors = error{
+    DBusConnectionRequired,
+    NoDBusUniqueName,
+    PutFailed,
+};
+
+pub fn init(
+    self: *OpenURI,
+    app: *App,
+) error{DBusConnectionRequired}!void {
+    const gio_app = app.app.as(gio.Application);
+    const dbus = gio_app.getDbusConnection() orelse {
+        return error.DBusConnectionRequired;
+    };
+
+    self.* = .{
+        .app = app,
+        .dbus = dbus,
+    };
+}
+
+/// Send the D-Bus method call to the XDG Desktop portal. The result of the
+/// method call will be reported asynchronously.
+pub fn start(self: *OpenURI, value: apprt.action.OpenUrl) (Allocator.Error || Errors)!void {
+    const alloc = self.app.core_app.alloc;
+
+    const token = portal.generateToken();
+
+    self.mutex.lock();
+    defer self.mutex.unlock();
+
+    try self.entries.putNoClobber(alloc, token, .{
+        .start = std.time.Instant.now() catch @panic("can't get the current time!"),
```

Comment:
I don't know how this possibly fails, but crashing a program for a failed URL open doesn't seem worth it. I'd prefer to see this turn into an error return (with proper cleanup if necessary).

Reply:
The only way `std.time.Instant.now()` fails if the system it's being run on lacks timers, which is AFAIK impossible on the sorts of systems that Ghostty runs on. So maybe just `unreachable` then.